70_FLIPFluids/main.cpp

```diff
@@ -0,0 +1,2240 @@
+		for (int i = 0; i < pressureSolverIterations; i++)
+		{
+			{
+				SMemoryBarrier memBarrier;
+				memBarrier.srcStageMask = PIPELINE_STAGE_FLAGS::COMPUTE_SHADER_BIT;
+				memBarrier.srcAccessMask = ACCESS_FLAGS::SHADER_WRITE_BITS;
+				memBarrier.dstStageMask = PIPELINE_STAGE_FLAGS::COMPUTE_SHADER_BIT;
+				memBarrier.dstAccessMask = ACCESS_FLAGS::SHADER_READ_BITS;
+				cmdbuf->pipelineBarrier(E_DEPENDENCY_FLAGS::EDF_NONE, {.memBarriers = {&memBarrier, 1}});
+			}
+
+			cmdbuf->bindDescriptorSets(nbl::asset::EPBP_COMPUTE, m_solvePressurePipeline->getLayout(), 1, 1, &m_solvePressureDs[i % 2].get());
+			cmdbuf->dispatch(WorkgroupCountGrid, 1, 1);
+		}
+
+		{
+			SMemoryBarrier memBarrier;
+			memBarrier.srcStageMask = PIPELINE_STAGE_FLAGS::COMPUTE_SHADER_BIT;
+			memBarrier.srcAccessMask = ACCESS_FLAGS::SHADER_WRITE_BITS;
+			memBarrier.dstStageMask = PIPELINE_STAGE_FLAGS::COMPUTE_SHADER_BIT;
+			memBarrier.dstAccessMask = ACCESS_FLAGS::SHADER_READ_BITS;
+			cmdbuf->pipelineBarrier(E_DEPENDENCY_FLAGS::EDF_NONE, {.memBarriers = {&memBarrier, 1}});
+        }
+
+		cmdbuf->bindComputePipeline(m_updateVelPsPipeline.get());
+		cmdbuf->bindDescriptorSets(nbl::asset::EPBP_COMPUTE, m_updateVelPsPipeline->getLayout(), 1, 1, &m_updateVelPsDs.get());
+		cmdbuf->dispatch(WorkgroupCountGrid, 1, 1);
+	}
+	
+	void dispatchExtrapolateVelocities(IGPUCommandBuffer* cmdbuf)
+	{
+		{
+			IGPUCommandBuffer::SPipelineBarrierDependencyInfo::image_barrier_t imageBarriers[2];
+			imageBarriers[0].barrier = {
+				.dep = {
+					.srcStageMask = PIPELINE_STAGE_FLAGS::COMPUTE_SHADER_BIT,
+					.srcAccessMask = ACCESS_FLAGS::SHADER_WRITE_BITS | ACCESS_FLAGS::SHADER_READ_BITS,
+					.dstStageMask = PIPELINE_STAGE_FLAGS::COMPUTE_SHADER_BIT,
+					.dstAccessMask = ACCESS_FLAGS::SHADER_READ_BITS
+			}
+			};
+			imageBarriers[0].image = velocityFieldImageView->getCreationParameters().image.get();
+			imageBarriers[0].subresourceRange = {
+				.aspectMask = IImage::EAF_COLOR_BIT,
+				.baseMipLevel = 0u,
+				.levelCount = 1u,
+				.baseArrayLayer = 0u,
+				.layerCount = 1u
+			};
+			imageBarriers[0].oldLayout = IImage::LAYOUT::GENERAL;
+			imageBarriers[0].newLayout = IImage::LAYOUT::READ_ONLY_OPTIMAL;
+			imageBarriers[1].barrier = {
+				.dep = {
+					.srcStageMask = PIPELINE_STAGE_FLAGS::COMPUTE_SHADER_BIT,
+					.srcAccessMask = ACCESS_FLAGS::SHADER_WRITE_BITS | ACCESS_FLAGS::SHADER_READ_BITS,
+					.dstStageMask = PIPELINE_STAGE_FLAGS::COMPUTE_SHADER_BIT,
+					.dstAccessMask = ACCESS_FLAGS::SHADER_READ_BITS
+			}
+			};
+			imageBarriers[1].image = prevVelocityFieldImageView->getCreationParameters().image.get();
+			imageBarriers[1].subresourceRange = {
+				.aspectMask = IImage::EAF_COLOR_BIT,
+				.baseMipLevel = 0u,
+				.levelCount = 1u,
+				.baseArrayLayer = 0u,
+				.layerCount = 1u
+			};
+			imageBarriers[1].oldLayout = IImage::LAYOUT::GENERAL;
+			imageBarriers[1].newLayout = IImage::LAYOUT::READ_ONLY_OPTIMAL;
```

Comment:
don't do layout transitions from GENERAL to READ_ONLY, layouts only truly matter for images that get used for renderpass attachments (they can keep some compression)